### ISSUE TITLE
"unresolved symbol __atomic_compare_exchange_16" with gcc/11.2/0 on Summit

### ISSUE THREAD

**Issue by jeanlucf22** [lanl/bml#627]
When trying to build with offload ON on Summit, I get the following error at link time:
unresolved symbol __atomic_compare_exchange_16

Build script:

#!/bin/bash
module load cmake
module load cuda
module load gcc/11.2.0
module load openblas

rm -r build
rm -r install

MY_PATH=$(pwd)

#get jsrun with full path
JSRUN=$(which jsrun)
echo ${JSRUN}

export CC=${CC:=gcc}
export FC=${FC:=gfortran}
export CXX=${CXX:=g++}
export BLAS_VENDOR=${BLAS_VENDOR:=OpenBLAS}
export BML_OPENMP=${BML_OPENMP:=yes}
export BML_OMP_OFFLOAD=${BML_OMP_OFFLOAD:=yes}
export INSTALL_DIR=${INSTALL_DIR:="${MY_PATH}/install"}
export BML_TESTING=${BML_TESTING:=yes}
export BML_COMPLEX=${BML_COMPLEX:=yes}
export CMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE:=Release}
export EXTRA_CFLAGS=${EXTRA_CFLAGS:="-fopenmp"}
export BML_CUSPARSE=${BML_CUSPARSE:=yes}
export BML_COMPLEX=${BML_COMPLEX:=no}

#use jsrun to run tests on a compute node
export BML_NONMPI_PRECOMMAND=${BML_NONMPI_PRECOMMAND:=${JSRUN}}
export BML_NONMPI_PRECOMMAND_ARGS=${BML_NONMPI_PRECOMMAND_ARGS:="-n1;-a1;-g1;-c7"}

./build.sh install

**Comment by nicolasbock** [comment]
As far as I know this symbol is part of `libatomic`. Maybe the linker needs `-latomic`?

**Comment by jeanlucf22** [comment]
@nicolasbock I tried that already, but no success: same error message.

$ nm /sw/summit/gcc/11.2.0-0/lib64/libatomic.a | grep __atomic_compare_exchange_16

0000000000000000 T __atomic_compare_exchange_16

Do the zeros mean some other library needs to be linked with?

**Comment by nicolasbock** [comment]
> Do the zeros mean some other library needs to be linked with?

I don't think so. It's possible that the order of the linker argument needs to change though.

**Comment by jeanlucf22** [comment]
Seems to work with:
export EXTRA_CFLAGS=${EXTRA_CFLAGS:="-fopenmp -latomic -foffload=-misa=sm_35"}
and some of the changes introduced in #626

**Comment by nicolasbock** [comment]
(I don't have access to a machine with a GPU so I can test with `offload`)

**Comment by jeanlucf22** [comment]
To me it looks like it is a platform specific issue (not needed on AMD GPU for instance) and I'd rather have '-latomic' in a platform specific build script than in CML

**Comment by nicolasbock** [comment]
Yes, that makes sense. Thanks for the analysis!
